src/test/java/org/broadinstitute/hellbender/utils/pileup/PileupElementUnitTest.java

```diff
@@ -179,13 +258,171 @@ public void testCreatePileupForReadAndOffset1(){
     }
 
     @Test
-    public void testCreatePileupForReadAndOffset2(){
+    public void testCreatePileupForReadAndOffset2() {
         final GATKRead read = ArtificialReadUtils.createArtificialRead("10M10D10M");
         final PileupElement pe10 = PileupElement.createPileupForReadAndOffset(read, 10);
         Assert.assertEquals(pe10.getOffset(), 10);
 
         final PileupElement pe15 = PileupElement.createPileupForReadAndOffset(read, 15);
         Assert.assertEquals(pe15.getOffset(), 15);
+    }
+
+    @Test
+    public void testIsImmediatelyAfter_insideMs() {
+        final GATKRead read = ArtificialReadUtils.createArtificialRead("10M10D10M");
+        final PileupElement pe = PileupElement.createPileupForReadAndOffset(read, 5);
+        Assert.assertFalse(pe.atStartOfCurrentCigar(), "atStartOfCurrentCigar");
+        Assert.assertFalse(pe.atEndOfCurrentCigar(), "atEndOfCurrentCigar");
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyAfter(op), "isImmediatelyAfter " + op);
+        }
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyBefore(op), "isImmediatelyBefore " + op);
+        }
+    }
+
+    @Test
+    public void testIsImmediatelyAfter_afterLastM() {
+        final GATKRead read = ArtificialReadUtils.createArtificialRead("10M10D10M");
+        final PileupElement pe = PileupElement.createPileupForReadAndOffset(read, 9);
+        Assert.assertFalse(pe.atStartOfCurrentCigar(), "atStartOfCurrentCigar");
+        Assert.assertTrue(pe.atEndOfCurrentCigar(), "atEndOfCurrentCigar");
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyAfter(op), "isImmediatelyAfter " + op);
+        }
+
+        Assert.assertTrue(pe.isImmediatelyBefore(D));
+        for (final CigarOperator op : CigarOperator.values()) {
+            if (op != D) {
+                Assert.assertFalse(pe.isImmediatelyBefore(op), "isImmediatelyBefore " + op);
+            }
+        }
+    }
+
+    @Test
+    public void testIsImmediatelyAfter_afterFirstD() {
+        final GATKRead read = ArtificialReadUtils.createArtificialRead("10M10D10M");
+        final PileupElement pe = PileupElement.createPileupForReadAndOffset(read, 10);
+        Assert.assertFalse(pe.atEndOfCurrentCigar(), "atEndOfCurrentCigar");
+        Assert.assertTrue(pe.isImmediatelyAfter(D));
+        for (final CigarOperator op : CigarOperator.values()) {
+            if (op != D) {
+                Assert.assertFalse(pe.isImmediatelyAfter(op), "isImmediatelyAfter " + op);
+            }
+        }
+
+        for (final CigarOperator op : CigarOperator.values()) {
+            Assert.assertFalse(pe.isImmediatelyBefore(op));
+        }
+    }
+
+    @Test
+    public void testIsImmediatelyAfter_insideDs() {
```

Comment:
This test case is not "inside" the deletion, is it? Isn't it inside the second M operator instead, since the deletion consumes no bases in the bases array?